File: Simulation/src/simulation/Checker.java
package simulation;

/**
 * Checker
 * 
 * Will manage the checking and validation of inputs 
 * prior to conversion to equivalent UTF value.
 * 
 * @author <author>
 *
 */
public class Checker {
	public Checker() {
		//null class
	}
	
	/**
	 * Checks and validates input value if allowed to be 
	 * converted to equivalent UTF value.
	 * 
	 * It also checks if the input value contains "U+", which it will automatically
	 * omit.
	 * 
	 * @param input Unicode value without any prefixes
	 * @return True if input is a valid Unicode value, False if otherwise.
	 */
	public String CheckInput(String input) {
		if(input.isEmpty())
			return null;
		input = input.replaceAll("\s+","");
		if(input.length()>2)
			if(input.substring(0,4).toUpperCase().equals("U+0X"))
				input = input.substring(4,input.length());
			else if(input.substring(0,2).toUpperCase().equals("U+") || input.substring(0,2).toUpperCase().equals("0X"))
				input = input.substring(2, input.length());
		long min = Long.parseLong("0000", 16), max = Long.parseLong("10FFFFF",16); //RANGE LIMIT FOR UNICODE VALUES
		
		//will only accept hex inputs via issuing an exception call
		try {
			long inputLong = Long.parseLong(input,16);
			if(inputLong < min || inputLong > max) //check if outside range of allowed hex values
				return null;
		}catch(NumberFormatException n) {
			return null;
		}
		return input;
	}
	
	/**
	 * Simply returns whether input Unicode is valid (true) or not (false).
	 * 
	 * Checks and validates input value if allowed to be 
	 * converted to equivalent UTF value.
	 * 
	 * It also checks if the input value contains "U+", which it will automatically
	 * omit.
	 * @param input Unicode value with("U+")/without prefix.
	 * @return True if a valid Unicode (after fixing) or not. 
	 */
	public boolean CheckInputBool (String input) {
		if(CheckInput(input) != null)
			return true;
		return false;
	}
}

File: Simulation_JS/js/Checker.js
/**
 * Checker
 * 
 * Will manage the checking and validation of inputs 
 * prior to conversion to equivalent UTF value.
 */

class Checker{
    constructor(){

    }
    /**
     * Checks and validates input value if allowed to be 
     * converted to equivalent UTF value.
     * @param {String} input Unicode value without any prefixes
     * @returns True if input is a valid Unicode value, False if otherwise.
     */
    CheckInput(input){
        var min = parseInt("0000", 16), max = parseInt("10FFFFF",16);
        /**
         * Will resort to Naive Approach 
         * since the initial implementation
         * of Regex does not work properly on all cases. 
         * 
         * If there are better solutions that does the same thing
         * albeit faster, please do add.
         * 
         * Simply check if it only contains the characters: ABCDEF0123456789 Non-Case Sensitive
         */
        var inputCaps = input.toUpperCase(); //just to simplify conditions (will only check ascii values of uppercase letters and numbers)
        if(inputCaps.length == 0)
            return null;
        if(inputCaps.length>2) 
            if(inputCaps.substring(0,4).match("U+0X"))
                inputCaps = inputCaps.substring(4,input.length);
            if(inputCaps.substring(0,2).match("U+") || inputCaps.substring(0,2).match("0X"))
                inputCaps = inputCaps.substring(2, inputCaps.length);
        for(var i = 0; i < inputCaps.length; i++){
            //65-70 = A-F & 48-57 = 0-9
            if((inputCaps.charCodeAt(i) < 48 || inputCaps.charCodeAt(i) > 57) && (inputCaps.charCodeAt(i) < 65 || inputCaps.charCodeAt(i) > 70))
                return null;
        }
        if(parseInt(inputCaps,16) < min || parseInt(inputCaps, 16) > max)
            return null;
        return inputCaps;
    }

    /**
     * Checks and validates input value if allowed to be 
     * converted to equivalent UTF value.
     * @param {String} input  Unicode value with("U+")/without prefix.
     * @returns True if a valid Unicode (after fixing) or not. 
     */
    CheckInputBool(input){
        if(this.CheckInput(input) != null)
            return true;
        return false;
    }
}

export {Checker};